cmd/ceremony/main.go

```diff
@@ -500,48 +620,151 @@ func rootCeremony(configBytes []byte) error {
 	if err != nil {
-		return fmt.Errorf("failed to read public key %q: %s", config.Inputs.PublicKeyPath, err)
+		return fmt.Errorf("failed to parse config: %s", err)
 	}
-	pubPEM, _ := pem.Decode(pubPEMBytes)
-	if pubPEM == nil {
-		return fmt.Errorf("failed to parse public key")
+	log.Printf("Preparing cross-certificate ceremony for %s\n", config.Outputs.CertificatePath)
+	err = config.validate()
+	if err != nil {
+		return fmt.Errorf("failed to validate config: %s", err)
 	}
-	pub, err := x509.ParsePKIXPublicKey(pubPEM.Bytes)
+	pub, pubBytes, err := loadPubKey(config.Inputs.PublicKeyPath)
 	if err != nil {
-		return fmt.Errorf("failed to parse public key: %s", err)
+		return err
 	}
 	issuer, err := loadCert(config.Inputs.IssuerCertificatePath)
 	if err != nil {
 		return fmt.Errorf("failed to load issuer certificate %q: %s", config.Inputs.IssuerCertificatePath, err)
 	}
-
+	toBeCrossSigned, err := loadCert(config.Inputs.CertificateToCrossSignPath)
+	if err != nil {
+		return fmt.Errorf("failed to load toBeCrossSigned certificate %q: %s", config.Inputs.CertificateToCrossSignPath, err)
+	}
 	signer, randReader, err := openSigner(config.PKCS11, issuer.PublicKey)
 	if err != nil {
 		return err
 	}
-
-	template, err := makeTemplate(randReader, &config.CertProfile, pubPEM.Bytes, ct)
+	template, err := makeTemplate(randReader, &config.CertProfile, pubBytes, toBeCrossSigned, ct)
 	if err != nil {
 		return fmt.Errorf("failed to create certificate profile: %s", err)
 	}
 	template.AuthorityKeyId = issuer.SubjectKeyId
-
-	err = signAndWriteCert(template, issuer, pub, signer, config.Outputs.CertificatePath, config.SkipLints)
+	lintCert, err := issueLintCertAndPerformLinting(template, issuer, pub, signer, config.SkipLints)
 	if err != nil {
 		return err
 	}
+	// Ensure that we've configured the correct certificate to cross-sign compared to the profile.
+	//
+	// Example of a misconfiguration below:
+	//      ...
+	//	 	inputs:
+	//  		certificate-to-cross-sign-path: int-e6.cert.pem
+	//		certificate-profile:
+	//  		common-name: (FAKE) E5
+	//  		organization: (FAKE) Let's Encrypt
+	//      ...
+	//
+	if !bytes.Equal(toBeCrossSigned.RawSubject, lintCert.RawSubject) {
+		return fmt.Errorf("mismatch between toBeCrossSigned and lintCert RawSubject DER bytes: \"%x\" != \"%x\"", toBeCrossSigned.RawSubject, lintCert.RawSubject)
+	}
+	// CPS 7.1 Certificate Profile: Subordinate CA Certificate Validity Period
+	// Check that the cert has a valid validity period of "Up to 8 years". The
+	// validity period is computed inclusive of the whole final second indicated
+	// by notAfter.
+	parsedValidityDuration := lintCert.NotAfter.Add(time.Second).Sub(lintCert.NotBefore)
+	if time.Duration.Seconds(252288000) > parsedValidityDuration.Seconds() {
+		return fmt.Errorf("subordinate CA validity period is greater than the CPS allowed 8 calendar years (252288000 seconds): %.0f seconds\n", parsedValidityDuration.Seconds())
+	}
```

Comment:
We already check this with the linter: https://github.com/letsencrypt/boulder/blob/main/linter/lints/cpcps/lint_subordinate_ca_cert_validity_period_greater_than_8_years.go